### PR TITLE
feat: tree drag and drop

### PR DESCRIPTION
**Pull request by js87zz**
### Please check if the PR fulfills these requirements
- [x] It's submitted to right branch according to our branching model
- [x] It's right issue type on title
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [ ] It does not introduce a breaking change or has description for the breaking change

### Description
we added move tree row functionality by dragging or `moveRow` API.

#### 1. D&D operation
```js
const grid = new tui.Grid({
  el: document.getElementById('grid'),
  data: treeData,
  draggable: true,
  treeColumnOptions: {
    name: 'name',
    useCascadingCheckbox: true
  },
  // ...
}
```
* move to an another index

* append to another node

#### 2. drag row header customizing
The user can customize the `draggable` row header column through existing `rowHeaders` option
```js
class DraggableRenderer {
  constructor(props) {
    const el = document.createElement('span');
    el.innerHTML = `😀`;
    el.style.cursor = 'pointer'
    this.el = el;
  }

  getElement() {
    return this.el;
  }

  render(props) {
    this.el.innerHTML = `😀`;
  }
}

const grid = new Grid({
  el: document.getElementById('grid'),
  data: treeData,
  rowHeaders: [
    {
      type: 'draggable',
      renderer: {
        type: DraggableRenderer
      }
    },
  ],
  bodyHeight: 500,
  draggable: true,
  treeColumnOptions: {
    name: 'name',
    useCascadingCheckbox: true
  },
  // ...
}
```

#### 3. `moveRow` API
We added the option in `moveRow` API for tree data.
The option is configured with `appended` property which has boolean value. This property for only tree data and deciedes whether the row is appended to other row as the child.
```ts
interface OptMoveRow {
  appended?: boolean
}
public moveRow(rowKey: RowKey, targetIndex: number, options: OptMoveRow): void;
```

#### 4. custom event(`dragStart`, `drag`, `drop`)
- `dragStart`: Drag to start the movement of the row (only occurs if the `dragable` option is enabled)
  - **rowKey**: dragging row's rowkey
  - **floatingRow**: floating row DOM element for customizing.
- `drag`: Dragging to move row (only occurs if the `dragable` option is enabled)
  - **rowKey**: dragging row's rowkey
  - **targetRowKey**: The rowKey to move
- `drop`: When the drag is over and the row movement is complete. (only occurs if the `dragable` option is enabled)
  - **rowKey**: dragging row's rowkey
  - **targetRowKey**: The rowKey to move

---
Thank you for your contribution to TOAST UI product. 🎉 😘 ✨

### CODE REVIEW

File: packages/toast-ui.grid/docs/en/custom-event.md [hunk 1]
```diff
@@ -119,6 +119,9 @@ grid.on('mousedown', function(ev) {
 - `scrollEnd` : When scrolling at the bottommost
 - `beforeChange`: Before one or more cells is changed
 - `afterChange`: After one or more cells is changed
+- `dragStart`: Drag to start the movement of the row (only occurs if the `dragableRow` option is enabled)
+- `drag`: Dragging to move row (only occurs if the `dragableRow` option is enabled)
+- `drop`: When the drag is over and the row movement is complete. (only occurs if the `dragableRow` option is enabled)
```
Comment (lja1018) [inline packages/toast-ui.grid/docs/en/custom-event.md:124]: `dragableRow` 옵션이 아직 수정 안된 것 같습니다

File: packages/toast-ui.grid/docs/ko/custom-event.md [hunk 2]
```diff
@@ -119,6 +119,9 @@ grid.on('mousedown', (ev) => {
 - `scrollEnd` : 스크롤 위치가 가장 하단에 도달한 경우
 - `beforeChange`: 하나 또는 여러 개의 셀 값이 변경되기 전
 - `afterChange`: 하나 또는 여러 개의 셀 값이 변경된 후
+- `dragStart`: 드래그하여 로우의 이동을 시작했을 때(`draggableRow` 옵션이 활성화된 경우만 발생)
+- `drag`: 드래그하여 로우를 이동하는 중(`draggableRow` 옵션이 활성화된 경우만 발생)
+- `drop`: 드래그가 끝나고 로우 이동을 완료하였을 때(`draggableRow` 옵션이 활성화된 경우만 발생)
```
Comment (lja1018) [inline packages/toast-ui.grid/docs/ko/custom-event.md:124]: 여기도 아직 반영되지 않은 것 같습니다.

File: packages/toast-ui.grid/src/dispatch/tree.ts [hunk 3]
```diff
@@ -391,3 +408,57 @@ function postUpdateAfterManipulation(store: Store, rowIndex: number, rows?: Row[
...
+  const { data, column, id } = store;
+  const { rawData } = data;
+  const targetRow = rawData[targetIndex];
+
+  if (!targetRow || isSorted(data) || isFiltered(data)) {
+    return;
+  }
+
+  const currentIndex = findIndexByRowKey(data, column, id, rowKey, false);
+  const row = rawData[currentIndex];
```
Comment (lja1018) [inline packages/toast-ui.grid/src/dispatch/tree.ts:427]: 조건문 아래로 내리는건 어떤가요?

File: packages/toast-ui.grid/src/dispatch/tree.ts [hunk 4]
```diff
@@ -391,3 +408,57 @@ function postUpdateAfterManipulation(store: Store, rowIndex: number, rows?: Row[
...
+
+    getDataManager(id).push('UPDATE', targetRow, true);
+    getDataManager(id).push('UPDATE', row, true);
+
+    if (options.appended) {
+      appendTreeRow(store, originRow, { parentRowKey: targetRow.rowKey });
+    } else {
+      const { parentRowKey } = targetRow._attributes.tree!;
+      const parentIndex = findIndexByRowKey(data, column, id, parentRowKey);
+      let offset = targetIndex;
```
Comment (lja1018) [inline packages/toast-ui.grid/src/dispatch/tree.ts:449]: `let offset;`만 해도 될 것 같습니다!

File: packages/toast-ui.grid/src/query/draggable.ts [hunk 5]
```diff
@@ -13,38 +14,87 @@ interface PosInfo {
...
+function createCells(cell: Element) {
+  const el = document.createElement('div');
+  el.className = cls('floating-cell');
+  el.style.width = window.getComputedStyle(cell).width;
 
-    for (let i = 0; i < cell.childNodes.length; i += 1) {
-      // the cell is not complex structure, so there is no the performance problem
-      el.appendChild(cell.childNodes[i].cloneNode(true));
-    }
+  for (let i = 0; i < cell.childNodes.length; i += 1) {
```
Comment (lja1018) [inline packages/toast-ui.grid/src/query/draggable.ts:39]: `cell.childNodes.length` 캐시해서 사용하는건 어떠신가요?

File: packages/toast-ui.grid/src/query/tree.ts [hunk 6]
```diff
@@ -114,7 +114,7 @@ export function isHidden(row: Row) {
 export function isLeaf(row: Row) {
```
Comment (lja1018) [inline packages/toast-ui.grid/src/query/tree.ts:114]: `{_attributes, _leaf}`만 받아서 사용하시는건 어떠신가요?

File: packages/toast-ui.grid/src/store/column.ts [hunk 7]
```diff
@@ -346,15 +349,19 @@ function createComplexColumnHeaders(
   });
 }
 
-function createDraggableRowHeader() {
+function createDraggableRowHeader(rowHeaderColumn: OptRowHeader | null) {
+  const renderer = isObject(rowHeaderColumn)
+    ? rowHeaderColumn.renderer
+    : { type: RowHeaderDraggableRenderer };
+
   const draggebleColumn: ColumnInfo = {
```
Comment (lja1018) [inline packages/toast-ui.grid/src/store/column.ts:357]: `draggable`의 오타같습니다.

File: packages/toast-ui.grid/src/view/bodyArea.tsx [hunk 8]
```diff
@@ -140,15 +180,30 @@ class BodyAreaComp extends Component<Props> {
     });
 
     if (draggableInfo) {
```
Comment (lja1018) [inline packages/toast-ui.grid/src/view/bodyArea.tsx:182]: 의견인데 `if (!draggbleInfo) { return; }`으로 depth를 낮추는건 어떠신가요?

File: packages/toast-ui.grid/src/view/bodyArea.tsx [hunk 9]
```diff
@@ -222,21 +277,55 @@ class BodyAreaComp extends Component<Props> {
...
+  private dropRow = () => {
+    const { hasTreeColumn } = this.props;
+    const { rowKey } = this.draggableInfo!;
+
+    if (this.movedIndexInfo) {
+      const { index, appended, rowKey: targetRowKey } = this.movedIndexInfo;
+
+      const gridEvent = new GridEvent({ rowKey, targetRowKey });
+      /**
+       * Occurs when droppring the row
```
Comment (lja1018) [inline packages/toast-ui.grid/src/view/bodyArea.tsx:289]: `dropping`의 오타네요

**Review by lja1018 — APPROVED** [review]
리뷰완료합니다.

File: packages/toast-ui.grid/src/view/bodyArea.tsx [hunk 10]
```diff
@@ -115,18 +128,45 @@ class BodyAreaComp extends Component<Props> {
...
+          this.movedIndexInfo = { index, rowKey: rowKeyToMove };
+        } else {
+          line.style.display = 'none';
+          this.movedIndexInfo = { index, appended: true, rowKey: rowKeyToMove };
+          this.props.dispatch('addRowClassName', rowKeyToMove, PARENT_CELL_CLASS);
+        }
+      } else {
+        this.movedIndexInfo = { index, rowKey: rowKeyToMove };
+        this.props.dispatch('moveRow', rowKey, index);
+      }
```
Comment (shiren) [inline packages/toast-ui.grid/src/view/bodyArea.tsx:169]: 요런 조건문은 쬐금 이해하기가 힘들 것 같아요.
함수로 뽑아서 각 조건에서 하는 작업을 설명하면 어떨까요?

**Review by shiren — APPROVED** [review]
수고하셨습니다.

File: packages/toast-ui.grid/src/view/bodyArea.tsx (hunk 8 above)
Comment (js87zz) [reply packages/toast-ui.grid/src/view/bodyArea.tsx:182]: depth가 많지 않아서 이거는 우선 유지할게요ㅎㅎ

File: packages/toast-ui.grid/src/dispatch/tree.ts [hunk 11]
```diff
@@ -391,3 +408,57 @@ function postUpdateAfterManipulation(store: Store, rowIndex: number, rows?: Row[
...
+  const moveToChild = some((childRow) => childRow.rowKey === targetRow.rowKey, childRows);
+
+  if (!moveToChild) {
+    removeTreeRow(store, rowKey);
+    const originRow = getOriginObject(row as Observable<Row>);
+
+    getDataManager(id).push('UPDATE', targetRow, true);
+    getDataManager(id).push('UPDATE', row, true);
+
+    if (options.appended) {
```
[the remark this answers is not in the document]
Comment (js87zz) [reply packages/toast-ui.grid/src/dispatch/tree.ts:444]: 디폴트 값은 `appended: false`이고 저 옵션 자체가 빈 객체로 넘어오는 일은 없습니다! 앞 단에서 설정하여 넘어옵니다.

File: packages/toast-ui.grid/src/view/bodyArea.tsx (hunk 10 above)
Comment (js87zz) [reply packages/toast-ui.grid/src/view/bodyArea.tsx:169]: 각 조건을 함수로 추출하기엔 더 복잡하고 가독성이 떨어질 것 같아, 트리쪽 처리 부분을 함수로 추출하고 주석 설명 추가하였습니다~
```js
if (props.hasTreeColumn) {
  // 함수로 분리
  this.setTreeMovedIndexInfo(rowKeyToMove, index, offsetTop, height);
} else {
  // move the row to next index
  this.movedIndexInfo = { index, rowKey: rowKeyToMove };
  this.props.dispatch('moveRow', rowKey, index);
}

private setTreeMovedIndexInfo(rowKey: RowKey, index: number, offsetTop: number, height: number) {
  // ...
  // display line border to mark the index to move
  if (Math.abs(height - offsetTop) < ADDITIONAL_RANGE) {
    // ...
  // show the background color to mark parent row
  } else {
    // ...
  }
}

```

**Pull request merged**